### PR TITLE
apply BC to adjoint equation rhs

### LINKED ISSUE #101: Fix applying boundary conditions of the adjoint problem
**General description of task**

To compute the gradient correctly of PDE-based systems. Adjoint BC are needed for solving the adjoint problem. We do have the adjoint BC correctly applied to the adjoint system matrix but not to the rhs vector. The goal of this issue is to apply the BC correctly. 

**Definition of Done (Feature/change)**
- [x] Apply BC to the RHS of the adjoin system

**Definition of Done (Mandatory)**
- N/A Documentation added (docstrings on all public methods/classes)
- [x] Unit tests added/updated (and passing!)
- [ ] Code reviewed and approved by 2 CUQI-maintainers
- N/A Online documentation is rendered correctly (esp. math) 

**Definition of Done (Optional)**
- N/A HowTo/tutorial added

### PR DESCRIPTION
**Pull request by amal-ghamdi**
closes #101 

Note regarding tests:
---------------------

The current gradient tests (before this PR) pass regardless of this update. So as part of this PR, the parameters of one of the tests (`test_gradient_poisson`) were updated so that the test fails when the adjoint BC are not applied correctly. 

Finite difference vs adjoint based gradient before the fix:
image

Finite difference vs adjoint based gradient after the fix:
image

Note regarding documentation:
-------------------------------
The documentation build failed initially for this PR. So a one-line fix was added to the file `docs.yml`.

### CODE REVIEW

File: cuqipy_fenics/pde.py [hunk 1]
```diff
@@ -481,6 +481,9 @@ def gradient_wrt_parameter(self, direction, *args, **kwargs):
                 "Gradient wrt parameter for PDE with observation operator not implemented")
 
         adjoint_rhs = -direction.vector()
+        # Apply the adjoint boundary conditions to the adjoint rhs
+        for bc in self._adjoint_dirichlet_bcs:
+            bc.apply(adjoint_rhs)
```
Comment (chaozg) [inline cuqipy_fenics/pde.py:486]: Hi @amal-ghamdi , is there an example with `adjoint_dirichlet_bcs` being used? I'm just curious how the code is like.

Comment (chaozg) [reply cuqipy_fenics/pde.py:486]: Did we meet an unexpected adjoint gradient at the boundary when we were testing the 1D MYULA example, just like first plot you pasted above? (my memory is not clear actually and I might be talking nonsense now...) Is this PR to fix that problem?

Comment (amal-ghamdi) [reply cuqipy_fenics/pde.py:486]: Yes exactly, the inference at the Drichelt boundary was sometimes off (in some runs just by a little like the right side of the last plot in https://cuqi-dtu.github.io/CUQI-Book/myula/). So I investigated more, comparing the FD gradient and the adjoint-based gradient, and found out the issue.

Also https://cuqi-dtu.github.io/CUQI-Book/myula/ provides an example of using the adjoint boundary conditions and we use adjoint bc in testing the gradients in the tests here (test_PDEModel.py).

Comment (amal-ghamdi) [reply cuqipy_fenics/pde.py:486]: this figure in particular (i am just putting it here before the content on the link change)
Image

Comment (chaozg) [inline cuqipy_fenics/pde.py:486]: Hi @amal-ghamdi , is there an example with `adjoint_dirichlet_bcs` being used? I'm just curious how the code is like.

**Review by chaozg — APPROVED** [review]
LGTM

**Review by jakobsj — APPROVED** [review]
